Report diagnostic errors on compilerOptions - #1457

eddyw opened the pull request:
[issues/1457](https://github.com/fuse-box/fuse-box/issues/1457)

- ensure typescript is installed. Checks on `index.js` since it's the earliest when TS is not used
  - ensures minimal version of TS is `3.0`. Okay? :P
- replace `ScriptTarget` enum with `ts.ScriptTarget`
- parse and validate `compilerOptions`
  - only done once before cached
  - format error|warning|message using TS formatter but displayed as standard FuseBox errors/warnings format
  - once parsed, `compilerOptions` is compatible with `ts.CompilerOptions` which replaces strings with equivalent enums' values. For instance, `jsx: "react"` will be `jsx: ts.JsxEmit.React`. This is helpful for not having mixed configuration (string values | number values)
  - explicitly unset/set `compilerOptions` used in `transpileModule` [transpile.ts#L26](https://github.com/Microsoft/TypeScript/blob/865b3e786277233585e1586edba52bf837b61b71/src/services/transpile.ts#L26) **Edit: removed unset, I realized some of the options could be needed in plugins :P**
    - currently has no effect since `transpileModule` does exactly that. However, it'd spend less time trying to fix and parse internally `compilerOptions`. Probably the time spend is insignificant but I also consider that `transpileModule` could be ported to FuseBox since it's a small function and we could ignore the part where it tries to fix `compilerOptions` internally all over again for each file transpiled.
  - displays warnings|messages|errors but only throws on error to prevents transpilation with broken `compilerOptions`. Rather than having the user figuring out why it doesn't work, the diagnostic messages show a good explanation for every broken property in the config

Notes:
- All current tests pass
- I haven't figured out how to write a test to check when `this.context.fatal` throws with `Error: Invalid 'compilerOptions' settings`, so I checked manually on the playground breaking settings. e.g: `{ target: 'asdasd', jsx: 'asdasd' }`

nchanged commented on src/tests/QuantumCSSplitting.test.ts:
```diff
@@ -229,7 +229,7 @@ export class QuantumCSSSplittingTest {
 				env.fileShouldExist("styles.css.map");
 				env.fileShouldExist("chunks/dcbf7349-ea9fe601.css");
 				env.fileShouldExist("chunks/ea9fe601.css.map");
-				env.fileShouldExist("chunks/c3eff8a7-ea9fe601.js");
```
what happened here? ;-)

eddyw replied:
Hash seems to be different. I changed it back, could you point me in the right direction as where to look?

nchanged replied:
But the tests are failing. Are you running it on windows?

eddyw replied:
Hmm, I realized why. If you are in `master` branch and run `QuantumCSSplitting.test.ts` but before in `File.js` on `transpileUsingTypescript` change it to:
```javascript
				const res = ts.transpileModule(this.contents, this.getTranspilationConfig());
				console.dir(res, { compact: false })
				console.dir(this.getTranspilationConfig())
				return res
```
You'll see this:
```javascript
{
  outputText: '"use strict";\nexports.__esModule = true;\n' +
    'require("./foobar.scss");\n//# ' +
    'sourceMappingURL=foobar.js.map',
  diagnostics: [
    {
      file: undefined,
      start: undefined,
      length: undefined,
      messageText: "Argument for '--target' option must " +
        "be: 'es3', 'es5', 'es6', 'es2015', " +
        "'es2016', 'es2017', 'es2018', " +
        "'esnext'.",
      category: 1,
      code: 6046,
      reportsUnnecessary: undefined
    }
  ],
  sourceMapText: '{"version":3,"file":"foobar.js","sourceRoot":"","sources":["foobar.ts"],"names":[],"mappings":";;AACE,yBAAsB","sourcesContent":["\\n\\t\\timport ' +
    '\\"./foobar.scss\\"\\n"]}'
}
{ compilerOptions:
   { module: 'commonjs',
     target: 'ES7',
     sourceMap: true,
     inlineSources: true,
     jsx: 'react',
     importHelpers: true,
     emitDecoratorMetadata: true,
     experimentalDecorators: true },
  fileName:
   '/Users/huangjx/Desktop/eddyw/Github/fuse-box/.fusebox/virtual-files/1548244391810/foobar.ts',
  transformers: {} }
```
That's because on default, we're using `ES7` .. but it's an invalid target. My branch fixes the issue.
`outputText` in `master` for last test:
```javascript
  outputText: '"use strict";\nexports.__esModule = true;\n' +
    'require("./foobar.scss");\n//# ' +
    'sourceMappingURL=foobar.js.map',
```
`outputText` in my branch:
```  outputText: '"use strict";\nObject.defineProperty(exports, ' +
    '"__esModule", { value: true });\n' +
    'require("./foobar.scss");\n//# ' +
    'sourceMappingURL=foobar.js.map',
```

eddyw replied:
I reverted last commit to which changes the hash on Quantum test, we were not transpiling to the correct target.
FuseBox default `compilerOptions` uses `{ target: "ES7" }` which should have been `{ target: "es2016" }`. This branch uses `ts.ScriptTarget` and parses and validates tsConfig, so compilerOptions passed are `{ target: 3 }` which corresponds to `es2016`|`es7`
[File.ts#L18](https://github.com/fuse-box/fuse-box/blob/master/src/core/File.ts#L18)

Since errors were not checked before in CompilerOptions, those were not caught. And since `transpileModule` always transpiles regardless of errors, the output now it wrong:
```javascript
exports.__esModule = true;
require("./foobar.scss");
```
Expected:
```javascript
Object.defineProperty(exports, "__esModule", { value: true });
require("./foobar.scss");```

eddyw replied:
Local simple test:
```const ts = require('ts');
const a = ts.transpileModule('export default 123', { compilerOptions: { module: 'commonjs', target: 'es2016' } }) // <<< expected
const b = ts.transpileModule('export default 123', { compilerOptions: { module: 'commonjs', target: 'ES7' } }) // <<< fails```

eddyw replied:
@nchanged ping 🙈

nchanged replied:
@eddyw sorry ;-) thanks for pinging. Will check it out today

eddyw replied:
@nchanged ping 🙈
Anyway to help?

nchanged replied:
There is some very annoying timing issue (racing condition) with this test which bugs me from time to time (

nchanged commented on src/core/TypescriptConfig.ts:
```diff
@@ -82,15 +105,98 @@ export class TypescriptConfig {
 		}
 	}
+
+	public categorizeDiagnostics(diagnostics: ReadonlyArray<ts.Diagnostic> | Readonly<ts.Diagnostic>): ICategorizedDiagnostics {
```
how can I test this in action?

eddyw replied:
In playground, modify `tsconfig.json` file with invalid settings such as `{ "target": "asd", "jsx": "asd" }`

nchanged commented on src/Utils.ts:
```diff
@@ -61,6 +62,40 @@ export function printCurrentVersion() {
+			console.log(NO_TYPESCRIPT_MESSAGE);
+			process.exit(1);
+		}
+		throw error;
```
Do you think we can make it more beautiful ?
<img width="909" alt="screenshot 2019-01-28 20 56 49" src="https://user-images.githubusercontent.com/1629088/51862821-bfd79800-233f-11e9-880d-e3991bc78bc6.png">

I think we can just take the first line from the stack trace and print it nicely. like we print that red message

nchanged commented:
@eddyw some small nitpick, but other than that it works flawlessly! Good job!

eddyw commented:
@nchanged 
- Prettier error messages
  - suggests where to look if tsconfig file was used/found or if set over-writes in `tsConfig` fuseBox option

![image](https://user-images.githubusercontent.com/1407526/51936296-d31f5c00-2442-11e9-9029-b72db3d630dd.png)

- docs suggest that `tsConfig` can be an array of compilerOptions, however it was typed only as `string`, I made some small changes to allow multiple arrays on `tsConfig` and typed the array for compilerOptions, so the user will know which options they can set in `tsConfig: [{ [CompilerOptions] },...]`
- pulled from master and merged into my branch
  - fixed conflicts, mostly in `ScriptTarget` uses

nchanged commented:
@eddyw https://github.com/fuse-box/fuse-box/commit/6c6c562b9abb37cb6eff9472c2df06a636898d1c

It seems that we all tried to fix the same issue here ;-)

nchanged commented:
@eddyw I can't really merge it, since there will be errors when a `tsconfig.json` file is generated. Could you drop the recent changes with targets, and we can see how we can improve ScriptTarget in a different PR?

eddyw commented:
@nchanged commented at the same time above haha
Yes, I noticed the change. However, in my branch I'm relying on `ts.ScriptTarget` which is more independent on TS version installed by the user. e.g: if we happen to have `ES2019` later, it'd still work working change and the check on compiler options would suggest all available targets.

eddyw commented:
@nchanged commented at the same time again :P
If I drop changes with targets, the errors may displayed about the `target` may not be accurate (according to TS version installed by user)

Maybe I could help improve when the `tsconfig` file is generated?

nchanged commented:
@eddyw yeah we do comment at the same time ;-) Of course, please read carefully my commit over there, (btw, there is console.log which needs to be removed) Make sure that an automatically generated `tsconfig.json` has the correct target (default is es2018 - I think it'll do just fine)

And squash yours commits + rebase. Github tells me that I cannot rebase.

nchanged commented:
@eddyw 
how about this condition (and in a few more places)

https://github.com/fuse-box/fuse-box/blob/6c6c562b9abb37cb6eff9472c2df06a636898d1c/src/core/File.ts#L207-L211

eddyw commented:
> @eddyw
> how about this condition (and in a few more places)
> 
> [fuse-box/src/core/File.ts](https://github.com/fuse-box/fuse-box/blob/6c6c562b9abb37cb6eff9472c2df06a636898d1c/src/core/File.ts#L207-L211)
> 
> Lines 207 to 211 in [6c6c562](/fuse-box/fuse-box/commit/6c6c562b9abb37cb6eff9472c2df06a636898d1c)
> 
>  public setLanguageLevel(level: ScriptTarget) { 
>  	if (this.languageLevel < level) { 
>  		this.languageLevel = level; 
>  	} 
>  }

It works the same way as when there was a custom `ScriptTarget`

eddyw commented:
@nchanged 
there is this function `normalizeAndValidateCompilerOptions` which is called to normalize compilerOptions, so if `target` is string (e.g: 'es5' or 'ES5') after analyze and normalize, we'd have `target: 1` (or equivalent enum numeric value always). Before FuseBox showed, for example `Typescript script target: 1` but I also changed that to display `ts.ScriptTarget[config.compilerOptions.target]` instead since `target` will be numeric always, we get the string value from enum, in the previous case `ES5`
https://github.com/fuse-box/fuse-box/blob/1f103c400199d19dc405a975ac252902c0ed5344/src/core/TypescriptConfig.ts#L89-L97
Btw, squashed and rebased some commits

nchanged commented:
@eddyw still can't rebase it ;-(

nchanged commented:
@eddyw there is a problem somewhere. FuseBox transpiles node modules using the latest preset ignoring the config.. (
